### PR TITLE
Telemetry events refactoring

### PR DESCRIPTION
**Pull request by antrix1989**

### CODE REVIEW

File: IdentityCore/src/telemetry/MSIDAggregatedDispatcher.m [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                                       MSID_TELEMETRY_KEY_CACHE_EVENT_COUNT: @(MSIDTelemetryCollectionBehaviorCollectAndCount),
+                                       };
+    }
+}
+
+#pragma mark - MSIDTelemetryDispatcher
+
+- (void)flush:(NSString *)requestId
+{
+    NSArray<id<MSIDTelemetryEventInterface>> *events = [self popEventsForReuquestId:requestId];
```
Comment (oldalton) [inline IdentityCore/src/telemetry/MSIDAggregatedDispatcher.m]: typo: popEventsForReuquestId

File: IdentityCore/src/telemetry/MSIDAggregatedDispatcher.m [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    
+    if (events.count == 0) return;;
```
Comment (oldalton) [inline IdentityCore/src/telemetry/MSIDAggregatedDispatcher.m]: nit: double ;

File: IdentityCore/src/telemetry/MSIDDefaultDispatcher.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+@protocol MSIDTelemetryEventsObserving;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface MSIDDefaultDispatcher : NSObject <MSIDTelemetryDispatcher>
+
+- (instancetype _Nullable)init NS_UNAVAILABLE;
++ (instancetype _Nullable)new NS_UNAVAILABLE;
+
+- (instancetype)initWithObserver:(id<MSIDTelemetryEventsObserving>)observer setTelemetryOnFailure:(BOOL)setTelemetryOnFailure;
```
Comment (oldalton) [inline IdentityCore/src/telemetry/MSIDDefaultDispatcher.h]: would it makes sense to call setTelemetryOnFailure something like "notifyOnFailureOnly"?

File: IdentityCore/src/telemetry/MSIDDefaultDispatcher.m [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        events = [self.eventsToBeDispatched[requestId] copy];
+        [self.eventsToBeDispatched removeObjectForKey:requestId];
+    });
+    
+    return events;
+}
+
+#pragma mark - Private
+
+- (NSDictionary *)appendPrefixForEvent:(NSDictionary *)event
```
Comment (oldalton) [inline IdentityCore/src/telemetry/MSIDDefaultDispatcher.m]: just to double check: this applies also for aggregated dispatcher?

File: IdentityCore/src/telemetry/MSIDTelemetryBaseEvent.m [hunk 5]
```diff
@@ -134,6 +134,21 @@ - (void)addDefaultProperties
...
 }
 
++ (NSArray<NSString *> *)propertiesToAggregate
+{
+    static dispatch_once_t once;
+    static NSArray *names = nil;
+    
+    dispatch_once(&once, ^{
+        names = @[
+                  MSID_TELEMETRY_KEY_REQUEST_ID,
```
Comment (oldalton) [inline IdentityCore/src/telemetry/MSIDTelemetryBaseEvent.m]: don't we also want to aggregate metadata events? e.g. x-client-ver etc

File: IdentityCore/src/telemetry/MSIDAggregatedDispatcher.m (hunk 1 above)
Comment (antrix1989) [reply IdentityCore/src/telemetry/MSIDAggregatedDispatcher.m]: done

File: IdentityCore/src/telemetry/MSIDAggregatedDispatcher.m (hunk 2 above)
Comment (antrix1989) [reply IdentityCore/src/telemetry/MSIDAggregatedDispatcher.m]: fixed

File: IdentityCore/src/telemetry/MSIDDefaultDispatcher.m (hunk 4 above)
Comment (antrix1989) [reply IdentityCore/src/telemetry/MSIDDefaultDispatcher.m]: yes

File: IdentityCore/src/telemetry/MSIDTelemetryBaseEvent.m (hunk 5 above)
Comment (antrix1989) [reply IdentityCore/src/telemetry/MSIDTelemetryBaseEvent.m]: Metadata events are always there, see [here](https://github.com/AzureAD/microsoft-authentication-library-common-for-objc/blob/20093b9fc7a9313ecf66dd3f601e15af7135f549/IdentityCore/src/telemetry/MSIDAggregatedDispatcher.m#L64).

File: IdentityCore/src/telemetry/MSIDDefaultDispatcher.h (hunk 3 above)
Comment (antrix1989) [reply IdentityCore/src/telemetry/MSIDDefaultDispatcher.h]: I like it, should we rename it in MSAL's public header too?

Comment (oldalton) [reply IdentityCore/src/telemetry/MSIDDefaultDispatcher.h]: If we rename dispatcher, I'm fine renaming this one as well.

Comment (antrix1989) [reply IdentityCore/src/telemetry/MSIDDefaultDispatcher.h]: renamed

**Pull request merged**